File: src/services/terminal.js
import { spawn } from 'child_process';
import { existsSync } from 'fs';
import { store } from '../store.js';

const sessions = new Map();
let nextId = 1;

// Detect best available shell on Windows: Git Bash > WSL > cmd.exe
// Git Bash preferred over WSL because it shares Windows' network stack —
// WSL has its own loopback, so 127.0.0.1 inside WSL doesn't reach the Windows host.
function detectWindowsShell() {
  const gitBashCandidates = [
    'C:\\Program Files\\Git\\bin\\bash.exe',
    'C:\\Program Files (x86)\\Git\\bin\\bash.exe',
    process.env.USERPROFILE ? `${process.env.USERPROFILE}\\AppData\\Local\\Programs\\Git\\bin\\bash.exe` : null,
  ].filter(Boolean);
  for (const p of gitBashCandidates) {
    if (existsSync(p)) return { type: 'gitbash', path: p };
  }

  const wslPath = `${process.env.SystemRoot || 'C:\\Windows'}\\System32\\wsl.exe`;
  if (existsSync(wslPath)) return { type: 'wsl', path: wslPath };

  return { type: 'cmd', path: process.env.ComSpec || 'cmd.exe' };
}

let _winShell = null;
function getWindowsShell() {
  return _winShell ??= detectWindowsShell();
}

const CLI_ENV = {
  AWS_DEFAULT_REGION:    'us-east-1',
  AWS_ENDPOINT_URL:      'http://localhost:4566',
  AWS_ACCESS_KEY_ID:     'local',
  AWS_SECRET_ACCESS_KEY: 'local',
};

export function createSession(type, instanceId) {
  if (type === 'ec2') {
    const inst = store.ec2.instances[instanceId];
    if (!inst) throw new Error(`Instance ${instanceId} not found`);
    if (!inst.containerId) throw new Error(
      'Instance has no Docker container — only Docker-backed instances support Connect.\n' +
      'Make sure Docker Desktop is running when you launch the instance.'
    );
  }

  const id = `sess-${nextId++}`;
  const session = {
    id, type,
    instanceId: instanceId || null,
    buffer: [], subs: new Set(),
    closed: false, busy: false,
    currentProc: null,
  };

  session.push = chunk => {
    session.buffer.push(chunk);
    if (session.buffer.length > 8000) session.buffer.shift();
    for (const fn of session.subs) fn(chunk);
  };

  // Welcome message — no process spawned yet
  if (type === 'cli') {
    let shellLabel;
    if (process.platform === 'win32') {
      const ws = getWindowsShell();
      shellLabel = ws.type === 'wsl' ? 'WSL (bash)' : ws.type === 'gitbash' ? 'Git Bash' : 'cmd.exe';
    } else {
      shellLabel = process.env.SHELL || '/bin/sh';
    }
    session.push({ t: 'o', d:
      '╔══════════════════════════════════════════════════════╗\r\n' +
      '║      MockCloud CLI  —  pre-configured shell        ║\r\n' +
      '╠══════════════════════════════════════════════════════╣\r\n' +
     `║  shell   ${shellLabel.slice(0,42).padEnd(44)}║\r\n` +
      '║  AWS_ENDPOINT_URL   = http://localhost:4566          ║\r\n' +
      '║  AWS_DEFAULT_REGION = us-east-1                     ║\r\n' +
      '╠══════════════════════════════════════════════════════╣\r\n' +
      '║  try:  aws s3 ls                                     ║\r\n' +
      '║        aws ec2 describe-instances                    ║\r\n' +
      '║        aws lambda list-functions                     ║\r\n' +
      '║        docker ps --filter label=mockcloud=ec2      ║\r\n' +
      '╚══════════════════════════════════════════════════════╝\r\n\r\n'
    });
  } else {
    const inst = store.ec2.instances[instanceId];
    session.push({ t: 'o', d:
      `Connected to ${inst.name || instanceId}\r\n` +
      `  OS:        ${inst.os}\r\n` +
      `  Type:      ${inst.type}\r\n` +
      `  Container: ${inst.containerId}\r\n\r\n`
    });
  }

  sessions.set(id, session);
  // Auto-expire after 30 min
  setTimeout(() => { if (sessions.has(id)) { closeSession(id); } }, 30 * 60 * 1000);
  return id;
}

export const getSession = id => sessions.get(id) ?? null;

export function subscribe(id, fn) {
  const s = sessions.get(id);
  if (!s) throw new Error('Session not found');
  s.subs.add(fn);
  return [...s.buffer];
}

export function unsubscribe(id, fn) {
  sessions.get(id)?.subs.delete(fn);
}

// Each command gets its own short-lived process — avoids TTY / buffering issues
export function execCommand(sessionId, command) {
  const s = sessions.get(sessionId);
  if (!s) throw new Error('Session not found');
  if (s.busy) throw new Error('A command is already running — press Ctrl+C to cancel');

  let cmd, args, env;

  if (s.type === 'ec2') {
    const inst = store.ec2.instances[s.instanceId];
    if (!inst?.containerId) {
      s.push({ t: 'e', d: 'Error: container not available\r\n' });
      s.push({ t: 'r', code: 1 });
      return;
    }
    cmd  = 'docker';
    args = ['exec', inst.containerId, '/bin/sh', '-c', command];
    env  = process.env;
  } else {
    if (process.platform === 'win32') {
      const ws = getWindowsShell();
      if (ws.type === 'wsl') {
        // Pass env vars inline since WSL doesn't inherit Windows env automatically
        const envPrefix = Object.entries(CLI_ENV).map(([k,v]) => `export ${k}=${v}`).join('; ');
        cmd  = ws.path;
        args = ['--', 'bash', '-c', `${envPrefix}; ${command}`];
        env  = process.env;
      } else if (ws.type === 'gitbash') {
        cmd  = ws.path;
        args = ['-c', command];
        env  = { ...process.env, ...CLI_ENV };
      } else {
        cmd  = ws.path;
        args = ['/d', '/c', command];
        env  = { ...process.env, ...CLI_ENV };
      }
    } else {
      cmd  = '/bin/sh';
      args = ['-c', command];
      env  = { ...process.env, ...CLI_ENV };
    }
  }

  s.busy = true;
  const proc = spawn(cmd, args, { env, stdio: ['pipe', 'pipe', 'pipe'], windowsHide: true });
  s.currentProc = proc;

  proc.stdout.on('data', d => s.push({ t: 'o', d: d.toString() }));
  proc.stderr.on('data', d => s.push({ t: 'e', d: d.toString() }));

  proc.on('close', code => {
    s.busy = false;
    s.currentProc = null;
    s.push({ t: 'r', code: code ?? 0 });
  });

  proc.on('error', err => {
    s.busy = false;
    s.currentProc = null;
    s.push({ t: 'e', d: `\r\nCould not run command: ${err.message}\r\n` });
    s.push({ t: 'r', code: 1 });
  });

  proc.stdin.end(); // no stdin for one-shot commands
}

export function interrupt(sessionId) {
  const s = sessions.get(sessionId);
  if (s?.currentProc) {
    try { s.currentProc.kill('SIGTERM'); } catch {}
    try { s.currentProc.kill('SIGKILL'); } catch {}
  }
}

export function closeSession(id) {
  const s = sessions.get(id);
  if (s) {
    if (s.currentProc) { try { s.currentProc.kill(); } catch {} }
    s.push({ t: 'x', d: '0' });
    sessions.delete(id);
  }
}

File: internal/services/terminal/terminal.go
// Package terminal — port of src/services/terminal.js: host-shell sessions
// with a ring buffer + SSE subscribers. Gated off by default (the route layer
// enforces MOCKCLOUD_ENABLE_TERMINAL + loopback). Sessions live here, not in
// the store — they're module-local in Node too.
package terminal

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"sync"
	"time"
)

// Chunk is one buffered/streamed event: t "o" (stdout) | "e" (stderr) |
// "r" (result, with Code) | "x" (closed).
type Chunk struct {
	T    string `json:"t"`
	D    string `json:"d,omitempty"`
	Code *int   `json:"code,omitempty"`
}

type Session struct {
	ID      string
	mu      sync.Mutex
	buffer  []Chunk
	subs    map[int]func(Chunk)
	nextSub int
	closed  bool
	busy    bool
	proc    *exec.Cmd
}

type Manager struct {
	mu     sync.Mutex
	byID   map[string]*Session
	nextID int
}

func NewManager() *Manager { return &Manager{byID: map[string]*Session{}, nextID: 1} }

var cliEnv = map[string]string{
	"AWS_DEFAULT_REGION":    "us-east-1",
	"AWS_ENDPOINT_URL":      "http://localhost:4566",
	"AWS_ACCESS_KEY_ID":     "local",
	"AWS_SECRET_ACCESS_KEY": "local",
}

type winShell struct {
	typ  string // gitbash | wsl | cmd
	path string
}

var (
	detectOnce   sync.Once
	cachedShell  winShell
)

func detectWindowsShell() winShell {
	detectOnce.Do(func() {
		candidates := []string{
			`C:\Program Files\Git\bin\bash.exe`,
			`C:\Program Files (x86)\Git\bin\bash.exe`,
		}
		if up := os.Getenv("USERPROFILE"); up != "" {
			candidates = append(candidates, up+`\AppData\Local\Programs\Git\bin\bash.exe`)
		}
		for _, p := range candidates {
			if _, err := os.Stat(p); err == nil {
				cachedShell = winShell{"gitbash", p}
				return
			}
		}
		sysRoot := os.Getenv("SystemRoot")
		if sysRoot == "" {
			sysRoot = `C:\Windows`
		}
		if wsl := filepath.Join(sysRoot, "System32", "wsl.exe"); fileExists(wsl) {
			cachedShell = winShell{"wsl", wsl}
			return
		}
		comspec := os.Getenv("ComSpec")
		if comspec == "" {
			comspec = "cmd.exe"
		}
		cachedShell = winShell{"cmd", comspec}
	})
	return cachedShell
}

func fileExists(p string) bool { _, err := os.Stat(p); return err == nil }

// CreateSession — only 'cli' is supported (EC2 Connect went away with Docker).
func (m *Manager) CreateSession(sessionType string) (string, error) {
	if sessionType != "cli" {
		return "", fmt.Errorf("Unsupported session type: %s — only 'cli' is available", sessionType)
	}
	m.mu.Lock()
	id := "sess-" + strconv.Itoa(m.nextID)
	m.nextID++
	s := &Session{ID: id, subs: map[int]func(Chunk){}}
	m.byID[id] = s
	m.mu.Unlock()

	shellLabel := "/bin/sh"
	if sh := os.Getenv("SHELL"); sh != "" {
		shellLabel = sh
	}
	if runtime.GOOS == "windows" {
		ws := detectWindowsShell()
		switch ws.typ {
		case "wsl":
			shellLabel = "WSL (bash)"
		case "gitbash":
			shellLabel = "Git Bash"
		default:
			shellLabel = "cmd.exe"
		}
	}
	label := shellLabel
	if len(label) > 42 {
		label = label[:42]
	}
	s.push(Chunk{T: "o", D: "╔══════════════════════════════════════════════════════╗\r\n" +
		"║      MockCloud CLI  —  pre-configured shell        ║\r\n" +
		"╠══════════════════════════════════════════════════════╣\r\n" +
		"║  shell   " + padEnd(label, 44) + "║\r\n" +
		"║  AWS_ENDPOINT_URL   = http://localhost:4566          ║\r\n" +
		"║  AWS_DEFAULT_REGION = us-east-1                     ║\r\n" +
		"╠══════════════════════════════════════════════════════╣\r\n" +
		"║  try:  aws s3 ls                                     ║\r\n" +
		"║        aws ec2 describe-instances                    ║\r\n" +
		"║        aws lambda list-functions                     ║\r\n" +
		"╚══════════════════════════════════════════════════════╝\r\n\r\n"})

	// Auto-expire after 30 min (unref'd in Node; a stray goroutine+timer here).
	t := time.AfterFunc(30*time.Minute, func() { m.CloseSession(id) })
	_ = t
	return id, nil
}

func padEnd(s string, n int) string {
	for len([]rune(s)) < n {
		s += " "
	}
	return s
}

func (m *Manager) Get(id string) *Session {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.byID[id]
}

func (s *Session) push(c Chunk) {
	s.mu.Lock()
	s.buffer = append(s.buffer, c)
	if len(s.buffer) > 8000 {
		s.buffer = s.buffer[1:]
	}
	subs := make([]func(Chunk), 0, len(s.subs))
	for _, fn := range s.subs {
		subs = append(subs, fn)
	}
	s.mu.Unlock()
	for _, fn := range subs {
		fn(c)
	}
}

// Subscribe registers a sink and returns the buffered backlog + an unsubscribe.
func (s *Session) Subscribe(fn func(Chunk)) ([]Chunk, func()) {
	s.mu.Lock()
	id := s.nextSub
	s.nextSub++
	s.subs[id] = fn
	backlog := append([]Chunk{}, s.buffer...)
	s.mu.Unlock()
	return backlog, func() {
		s.mu.Lock()
		delete(s.subs, id)
		s.mu.Unlock()
	}
}

func (s *Session) Closed() bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.closed
}

// ExecCommand runs one short-lived process per command (avoids TTY/buffering
// issues), streaming stdout/stderr to subscribers.
func (m *Manager) ExecCommand(sessionID, command string) error {
	s := m.Get(sessionID)
	if s == nil {
		return fmt.Errorf("Session not found")
	}
	s.mu.Lock()
	if s.busy {
		s.mu.Unlock()
		return fmt.Errorf("A command is already running — press Ctrl+C to cancel")
	}
	s.busy = true
	s.mu.Unlock()

	name, args, env := buildCommand(command)
	proc := exec.Command(name, args...)
	proc.Env = env
	hideWindow(proc) // Node: windowsHide: true
	stdout, _ := proc.StdoutPipe()
	stderr, _ := proc.StderrPipe()
	s.mu.Lock()
	s.proc = proc
	s.mu.Unlock()

	if err := proc.Start(); err != nil {
		s.finish(func() { s.busy = false; s.proc = nil })
		s.push(Chunk{T: "e", D: "\r\nCould not run command: " + err.Error() + "\r\n"})
		s.push(Chunk{T: "r", Code: intp(1)})
		return nil
	}
	go func() {
		buf := make([]byte, 4096)
		for {
			n, err := stdout.Read(buf)
			if n > 0 {
				s.push(Chunk{T: "o", D: string(buf[:n])})
			}
			if err != nil {
				return
			}
		}
	}()
	go func() {
		buf := make([]byte, 4096)
		for {
			n, err := stderr.Read(buf)
			if n > 0 {
				s.push(Chunk{T: "e", D: string(buf[:n])})
			}
			if err != nil {
				return
			}
		}
	}()
	go func() {
		err := proc.Wait()
		code := 0
		if err != nil {
			if exitErr, ok := err.(*exec.ExitError); ok {
				code = exitErr.ExitCode()
			} else {
				code = 1
			}
		}
		s.finish(func() { s.busy = false; s.proc = nil })
		s.push(Chunk{T: "r", Code: &code})
	}()
	return nil
}

func (s *Session) finish(fn func()) {
	s.mu.Lock()
	fn()
	s.mu.Unlock()
}

func buildCommand(command string) (string, []string, []string) {
	if runtime.GOOS == "windows" {
		ws := detectWindowsShell()
		switch ws.typ {
		case "wsl":
			var prefix []string
			for k, v := range cliEnv {
				prefix = append(prefix, "export "+k+"="+v)
			}
			return ws.path, []string{"--", "bash", "-c", strings.Join(prefix, "; ") + "; " + command}, os.Environ()
		case "gitbash":
			return ws.path, []string{"-c", command}, mergedEnv()
		default:
			return ws.path, []string{"/d", "/c", command}, mergedEnv()
		}
	}
	return "/bin/sh", []string{"-c", command}, mergedEnv()
}

func mergedEnv() []string {
	env := os.Environ()
	for k, v := range cliEnv {
		env = append(env, k+"="+v)
	}
	return env
}

func (m *Manager) Interrupt(sessionID string) {
	s := m.Get(sessionID)
	if s == nil {
		return
	}
	s.mu.Lock()
	proc := s.proc
	s.mu.Unlock()
	if proc != nil && proc.Process != nil {
		_ = proc.Process.Kill()
	}
}

func (m *Manager) CloseSession(id string) {
	m.mu.Lock()
	s := m.byID[id]
	delete(m.byID, id)
	m.mu.Unlock()
	if s == nil {
		return
	}
	s.mu.Lock()
	proc := s.proc
	s.closed = true
	s.mu.Unlock()
	if proc != nil && proc.Process != nil {
		_ = proc.Process.Kill()
	}
	s.push(Chunk{T: "x", D: "0"})
}

func intp(n int) *int { return &n }
